From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Juhyung Park <qkrwngud825@gmail.com>
Date: Thu, 11 Apr 2024 21:38:53 +0900
Subject: [PATCH 30/34] libc: allow selectively enabling jemalloc's
 zero-filling

Scudo's nature of being a hardened memory allocator meaned that
improper memory accesses are catched more strictly and vendors had to
fix improper memory accesses from their binaries. This used to be a
problem when switching from jemalloc to Scudo, but now, the inverse has
occured: switching from Scudo to jemalloc is crashing some vendor
binaries.

This is mostly due to the fact that Scudo returns sanitized buffer
while jemalloc doesn't and vendor was lazy enough to not handle
uninitialized buffer properly.

As we do not have the source code to fix such bad binaries, add an
environment interface to selectively enable jemalloc's zero-filling.

To make use of this, add "setenv JE_MALLOC_ZERO_FILLING 1" to the
problematic service's rc file.

Change-Id: Ib3428815a9f5da3787a19f8a18f7609be747659b
Signed-off-by: Juhyung Park <qkrwngud825@gmail.com>
---
 libc/bionic/malloc_common_dynamic.cpp | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/libc/bionic/malloc_common_dynamic.cpp b/libc/bionic/malloc_common_dynamic.cpp
index e2c6eb1..b321d3d 100644
--- a/libc/bionic/malloc_common_dynamic.cpp
+++ b/libc/bionic/malloc_common_dynamic.cpp
@@ -368,11 +368,15 @@ static bool InstallHooks(libc_globals* globals, const char* options, const char*
   return true;
 }
 
+#if defined(USE_SCUDO)
 extern "C" const char* __scudo_get_stack_depot_addr();
 extern "C" const char* __scudo_get_region_info_addr();
 extern "C" const char* __scudo_get_ring_buffer_addr();
 extern "C" size_t __scudo_get_ring_buffer_size();
 extern "C" size_t __scudo_get_stack_depot_size();
+#else
+extern "C" void je_set_zero_filling(bool val);
+#endif
 
 // Initializes memory allocation framework once per process.
 void MallocInitImpl(libc_globals* globals) {
@@ -381,13 +385,22 @@ void MallocInitImpl(libc_globals* globals) {
 
   MaybeInitGwpAsanFromLibc(globals);
 
-#if defined(USE_SCUDO) && !__has_feature(hwaddress_sanitizer)
+#if defined(USE_SCUDO)
+#if !__has_feature(hwaddress_sanitizer)
   __libc_shared_globals()->scudo_stack_depot = __scudo_get_stack_depot_addr();
   __libc_shared_globals()->scudo_region_info = __scudo_get_region_info_addr();
   __libc_shared_globals()->scudo_ring_buffer = __scudo_get_ring_buffer_addr();
   __libc_shared_globals()->scudo_ring_buffer_size = __scudo_get_ring_buffer_size();
   __libc_shared_globals()->scudo_stack_depot_size = __scudo_get_stack_depot_size();
 #endif
+#else
+  // Check if the process wants zero-filling
+  char* env = getenv("JE_MALLOC_ZERO_FILLING");
+  if (env && strcmp(env, "1") == 0) {
+    async_safe_format_log(ANDROID_LOG_WARN, "libc", "Enabling jemalloc zero-filling");
+    je_set_zero_filling(true);
+  }
+#endif
 
   // Prefer malloc debug since it existed first and is a more complete
   // malloc interceptor than the hooks.
